### PR TITLE
Repo Refactoring and Cleanup

### PR DESCRIPTION
**Pull request by asirbu77**
### 1. Build & Dependency Management
- **Added uv support** with updated `pyproject.toml`
- **CUDA 12 and CUDA 13 support** via optional dependencies
- Uses developer version of MuJoCo and overnight build of MJX for `mujoco_playground` compatibility
- Updated `environment.yml` and dependency configurations
- Installs `vnl-playground` from tagged release (previously editable mode)

### 2. Entry Point & Training Pipeline
- **Renamed `train_vnl.py` → `train.py`** as the new main entry point
- Cleaned up training script with improved structure
- Added train/test splitting logic for data handling
- Switched optimizer from Muon to AdamW (better with soft resets)

### 3. Configuration System
- **New config:** `rodent-full-clips.yaml` (updated from legacy config)
- **Deleted:** `mouse-arm.yaml` (deprecated)
- Added `track_mjx/config/utils.py` with configuration utility functions
- Added detailed comments to config files
- Deprecated config conversion function

### 4. Checkpointing & Preemption
- **Consolidated `preemption.py` logic into `checkpointing.py`**
- Added checkpointing utility functions
- Fixed checkpointing bugs
- Enhanced checkpoint loading and rollout capabilities

### 5. Logging & Output
- Converted `print` statements to `logging.info`
- Added exception handling to WandB rendering
- Improved WandB logging with higher resolution rendering
- Uses vnl-playground rendering functions for WandB

---

### LSTM Support (Removed)
- `track_mjx/agent/lstm_ppo/` - entire directory removed:
  - `acting.py`
  - `intention_network.py`
  - `losses.py`
  - `ppo.py`
  - `ppo_networks.py`

### Legacy Environments (Removed)
- `track_mjx/environment/` - entire module removed:
  - Reacher environment (mouse arm assets, XML, code)
  - Walker environment (fruitfly assets)
  - Task definitions (single/multi clip tracking/reaching)
  - Reward functions

### Deprecated Files
- `track_mjx/agent/preemption.py` - logic moved to `checkpointing.py`
- `docs/installation.md` - removed

---

### Agent Module (`track_mjx/agent/`)

| File | Changes |
|------|---------|
| `checkpointing.py` | +534/-195 - Major rewrite with preemption logic |
| `wandb_logging.py` | +189/-147 - vnl-playground rendering integration |
| `network_masks.py` | +37/-6 - Multi-walker constants support |
| `mlp_ppo/intention_network.py` | +133/-40 - Cleaned with docstrings |
| `mlp_ppo/losses.py` | +107/-67 - Cleaned with type annotations |
| `mlp_ppo/ppo.py` | +109/-86 - Cleaned |
| `mlp_ppo/ppo_networks.py` | +141/-62 - Cleaned |

### Analysis Module (`track_mjx/analysis/`)

| File | Changes |
|------|---------|
| `render.py` | +134/-348 - Simplified, uses vnl-playground |
| `rollout.py` | +123/-307 - Generates vnl-playground compatible rollouts |
| `utils.py` | +82/-39 - Cleaned with docstrings |

### Training Parameters
- Changed `root_quat` exponential scale for stable training
- Updated `render_interval` to 1
- Added `max_start_frame` parameter (previously hardcoded)

## Summary

This PR represents a major refactoring effort that:

1. **Simplifies** the codebase by removing deprecated LSTM and environment code
2. **Modernizes** the build system with uv and multi-CUDA support
3. **Integrates** with vnl-playground for rendering and environment handling
4. **Improves** code quality with docstrings, type annotations, and better logging
5. **Consolidates** checkpointing/preemption into a unified module

### CODE REVIEW

File: notebooks/rollout_from_checkpoint.ipynb
Comment (charles-zhng) [inline notebooks/rollout_from_checkpoint.ipynb]: Don't need to push any changes to this notebook

File: track_mjx/agent/mlp_ppo/ppo.py [hunk 1]
```diff
@@ -27,6 +27,9 @@
 from brax.training import acting
 from brax.training import pmap
 from brax.training import types
+
+# from brax.training import gradients
```
Comment (charles-zhng) [inline track_mjx/agent/mlp_ppo/ppo.py]: remove comment

File: track_mjx/config/rodent-full-clips-vnl.yaml [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  ctrl_dt: 0.01  # sim_dt * physics_steps_per_control_step (0.002 * 5)
+  mocap_hz: 50
+  
+  # Reference clip settings
+  clip_length: 250
+  clip_set: "all"
+  reference_length: 5  # traj_length
+  start_frame_range: [0, 50]  # [0, random_init_range]
+  qvel_init: "zeros"
+  keep_clips_idx: null
```
Comment (charles-zhng) [inline track_mjx/config/rodent-full-clips-vnl.yaml]: null -> None?

File: .gitignore
Comment (charles-zhng) [inline .gitignore]: Don't need to push these changes

File: track_mjx/agent/preemption.py [hunk 3]
```diff
@@ -251,3 +253,82 @@ def checkpoint_callback(step: int):
             logging.warning(f"Failed to update run state after checkpoint save: {e}")
 
     return checkpoint_callback
+
+def load_from_run_state(cfg: DictConfig) -> Tuple[str, str, Optional[Dict[str, Any]]]:
```
[the remark this answers is not in the document]
Comment (asirbu77) [reply track_mjx/agent/preemption.py]: Yes. `preemption.py` is being deprecated. All its logic is moved to `checkpointing.py`

[the remark this answers is not in the document]
Comment (charles-zhng) [reply track_mjx/agent/preemption.py]: this file be removed then?

File: track_mjx/config/rodent-full-clips-vnl.yaml [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    - summed_pos_distance
+    - joint_distance
+    - quat_distance
+    - ctrl_cost
+    - ctrl_diff_cost
+    - energy_cost
+    - too_far
+    - bad_pose
+    - bad_quat
+    - fall
```
[the remark this answers is not in the document]
Comment (charles-zhng) [reply track_mjx/config/rodent-full-clips-vnl.yaml]: Why don't we just log all metrics? Ideally we dont have this long list in the config (esp if theyre not really configurable since it's a finite list defined by the env). The user can just filter in wandb anyway

File: track_mjx/utils.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    # Create ml_collections env_config
+    env_cfg_dict = OmegaConf.to_container(env_cfg, resolve=True)
+    env_cfg_ml = config_dict.ConfigDict(env_cfg_dict)
+
+    # Convert config to dict
+    logging.info(f"Configs: {OmegaConf.to_container(cfg, resolve=True)}")
+    cfg_dict = OmegaConf.to_container(cfg, resolve=True)
+
+    return (cfg, cfg_dict, env_cfg, env_cfg_ml, render_cfg, network_cfg, train_setup, train_cfg, logging_cfg, walker_cfg)
```
Comment (charles-zhng) [inline track_mjx/utils.py]: This is a bit crazy, I don't think we need to unpack this here

File: track_mjx/utils.py
Comment (charles-zhng) [inline track_mjx/utils.py]: maybe move this into `./config`?
